### ISSUE TITLE
Memory leak while using TLS on Mac OSX

### ISSUE THREAD

**Issue by dradhamohan** [dotnet/runtime#34080]
Hi,
I am using .NetCore 3.1.2 on Mac OS X. Have a simple program that just establishes a TLS connection with a server and closes immediately. Doing this repeatedly, say for every X milliseconds, we can observe that the process memory utilization increases monotonically. The same program runs on Windows platform without any issues. The issue is seen on High Sierra, Mojave and Catalina. 
Test program below. Just hoping that there isn't any silly mistake in the program and that I am not raising a false positive here.
Would appreciate if someone can point out the mistake in the program (or) acknowledge that this is really an issue in .NetCore OS X implementation of TLS stack. 
Thanks in advance !

Regards,
Radha

```c#
using System;
using System.Net.Sockets;
using System.Net.Security;
using System.Security.Authentication;
using System.Security.Cryptography.X509Certificates;
using System.Threading;

namespace TestSsl
{
    class Program
    {
        // The following method is invoked by the RemoteCertificateValidationDelegate.
        public static bool ValidateServerCertificate(
              object sender,
              X509Certificate certificate,
              X509Chain chain,
              SslPolicyErrors sslPolicyErrors)
        {

            Console.WriteLine("Check for server ceritificate");
            Console.WriteLine("{0}Subject: {1}{0}", Environment.NewLine, certificate.Subject);
            Console.WriteLine("{0}Issuer: {1}{0}", Environment.NewLine, certificate.Issuer);
            Console.WriteLine("{0}Certificate to string: {1}{0}", Environment.NewLine, certificate.ToString(true));
            if (sslPolicyErrors == SslPolicyErrors.None)
                return true;

            Console.WriteLine("Certificate error: {0}", sslPolicyErrors);

            // Do not allow this client to communicate with unauthenticated servers.
            return false;
        }



        public static void RunClient()
        {
            TcpClient client = new TcpClient("mytestdomain.com", 443); //its a local origin and shouldn't fail
            Console.WriteLine("Connection established."); 

            SslStream sslStream = new SslStream(innerStream: client.GetStream(), leaveInnerStreamOpen: true, userCertificateValidationCallback: (a, b, c, d) => true, userCertificateSelectionCallback: null);
            //SslStream sslStream = new SslStream(innerStream: client.GetStream(), leaveInnerStreamOpen: false, userCertificateValidationCallback: new RemoteCertificateValidationCallback(ValidateServerCertificate), userCertificateSelectionCallback: null);
            try
            {
                sslStream.AuthenticateAsClient("mytestdomain.com", null, SslProtocols.Tls12, false);
            }
            catch(Exception e)
            {
                Console.WriteLine("Exception raised " + e);
            }
            sslStream.Dispose();
            sslStream.Close();

            client.Dispose();
            client.Close();
            Console.WriteLine("Connection closed.");
        }
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
            int counter = 0;
            while(true)
            {
                RunClient();
                counter++;
                if ((counter % 100) == 0)
                {
                    GC.Collect();
                    counter = 0;
                }
                Thread.Sleep(20);
            }
        }
    }
}

```

**Comment by wfurt** [comment]
I will take a look. It should be reasonably easy to see if this is deferred collection or of there is real like - that would be probably in the PAL layer that is quite distinct on OSX.

**Comment by wfurt** [comment]
I can see process grow. I don't know if the Close after Dispose is intentional but changing that standard `using() {}` patter did not changed that. 

The managed heap goes up & down.


This is output collected after running for.a while with (dump heap -stat)
The `AppleCertificatePal` certainly looks suspicious.  I will take a look.


```
000000012580f888       95         9120 System.Net.Security.SslAuthenticationOptions
0000000125801108       95         9120 System.Net.Sockets.Socket
0000000125180be8      404         9696 System.Object
0000000125a67f38        1         9908 System.Collections.Generic.Dictionary`2+Entry[[System.Net.Security.TlsCipherSuite, System.Net.Security],[System.Net.Security.TlsCipherSuiteData, System.Net.Security]][]
0000000125a83680      192        10688 System.Runtime.InteropServices.SafeHandle[]
0000000125a83770      380        12160 Microsoft.Win32.SafeHandles.SafeCreateHandle
0000000125a6eba0      380        12160 System.Security.Cryptography.X509Certificates.SafeSecIdentityHandle
0000000125a6eb08      380        12160 System.Security.Cryptography.X509Certificates.SafeSecCertificateHandle
0000000125a86fb8      390        12480 Microsoft.Win32.SafeHandles.SafeCFDataHandle
00000001259042b8      285        13680 System.Net.Security.ProtocolToken
000000012580cf50       95        13680 System.Net.Sockets.SocketAsyncContext
0000000125a88188      380        15200 System.Security.Cryptography.X509Certificates.X500DistinguishedName
0000000125a7fae8      187        16456 System.TimeZoneInfo+AdjustmentRule
0000000125a8e518      191        16744 System.ValueTuple`3[[System.Security.Cryptography.Asn1.Asn1Tag, System.Security.Cryptography.X509Certificates],[System.Int32, System.Private.CoreLib],[System.Security.Cryptography.Asn1.UniversalTagNumber, System.Security.Cryptography.X509Certificates]][]
00000001254e4860       95        19000 System.Net.Security.SslStream
0000000125a6ed78      760        24320 System.Security.Cryptography.Apple.SafeKeychainHandle
0000000125ba23c8      381        24344 System.Security.Cryptography.X509Certificates.X509ChainStatus[]
0000000125a79658      571        27408 System.Text.StringBuilder
00000001252407c8       28        47812 System.Int32[]
0000000125a81ab0     1332        53280 System.Security.Cryptography.Oid
000000012540d340      573        55366 System.Char[]
0000000125231c10      966        58272 System.SByte[]
000000012590e578      380        60800 System.Security.Cryptography.X509Certificates.X509Certificate2
0000000125186628      216        62272 System.Object[]
0000000125405558     3422        95844 System.UInt32[]
0000000125236ed8     1009        98006 System.String
0000000125a6cce8      380       158080 Internal.Cryptography.Pal.AppleCertificatePal
0000000125299f80     3149      1145915 System.Byte[]
Total 23680 objects
```

cc: @bartonjs

**Comment by dradhamohan** [comment]
Do we understand that this is a .NetCore issue (and not an issue with the program) ? Can you please confirm ?

**Comment by alnikola** [comment]
Triage:
Looks like a PAL layer or native crypto library issue.

**Comment by wfurt** [comment]
I did few variations of the code and I don't see any hints that is your code @dradhamohan.

**Comment by dradhamohan** [comment]
Hi - Looks like .NetCore on Mac is using Apple Transport Layer Security APIs. Attaching a sample program that just repeatedly performs a TLS handshake and close. Interestingly, there seems to be a small memory increase for every handshake/close. Can this be an with the Apple TLS stack ? Or is there anything wrong in the Apple TLS API usage in the attached program ?
---------------snip------------
//
//  main.m
/*  
* tlsHandshakeTest - program to test memory leak during TLS handshake
* A simple program that wakes up every millisecond, performs a TLS handshake 
* with a TLS server (google.com) and closes the connection immediately (without any data transfer).
* Running this program on Mac High-Sierra/Mojave/Catalina would show a small and steady monotonic increase in the memory usage of the process (roughly 2KB increase per handshake).
* 
* Note: This program uses Apple Transport Layer Security APIs which are deprecated on Mojave and Catalina.
* 
*/
//

#import <Foundation/Foundation.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

typedef struct conn_s {
    int sock_fd;
    int want_read;
    int want_write;
}conn_t;

//I/O callback APIs
OSStatus readTLSConn(SSLConnectionRef connection, void *data, size_t *dataLength);
OSStatus writeTLSConn(SSLConnectionRef connection, const void *data, size_t *dataLength);

/*
 * Function: readTLSConn
 * Read I/O callback implementation.Reads data from the underying TCP socket.
 * @param data - buffer allocated by caller
 * @param dataLength - in/out parameter; contains the size of the input buffer. 
 * On successful read contains the number of bytes read from the underlying socket.
 */
OSStatus readTLSConn(SSLConnectionRef connection, void *data, size_t *dataLength)
{
    //NSLog(@"readTLSConn called...\n");
    conn_t *tls_conn = (conn_t*) connection;
    int sock_fd = tls_conn->sock_fd;
    size_t bytesRequested = *dataLength;
    //CFStringRef errString;
    *dataLength = 0;
    //NSLog(@"Bytes requested = %lu socket = %d", bytesRequested, sock_fd);
    
    ssize_t ret = 0;
    OSStatus status = noErr;
    tls_conn->want_read = 0;
    while (1) {
        ret = read(sock_fd, data, bytesRequested);
        if (ret > 0) {
            *dataLength += ret;
            if (bytesRequested > ret)
            {
                data = data + ret;
                bytesRequested -= ret;
            }
            else
            {
                status = errSecSuccess;
                break;
            }
        }
        else if (ret == 0)
        {
            status = errSSLClosedGraceful;
            break;
        }
        else
        {
            int error_var = 0;
            socklen_t size_of_error = sizeof(error_var);
            getsockopt(sock_fd, SOL_SOCKET, SO_ERROR, (char*)&error_var, &size_of_error);
            //NSLog(@"getsockopt error = %d", error_var);
            switch (errno) {
              case ENOENT:
                    status = errSSLClosedGraceful;
                    break;
              case EAGAIN:
                    tls_conn->want_read = 1;
                    //NSLog(@"Setting status as would block");
                    status = errSSLWouldBlock;
                    break;
              case ECONNRESET:
                    status = errSSLClosedAbort;
                    break;
              default:
                    status = errSecIO;
                    break;
            }
            break;
        }
    }
    //errString = SecCopyErrorMessageString(status, NULL);
    //NSLog(@"Number of bytes read = %d status = %@ errno = %d", *dataLength, errString, errno);
    return status;
}

/*
 * Function: writeTLSConn
 * Write I/O callback implementation.Reads data from the underying TCP socket.
 * @param data - buffer allocated by caller
 * @param dataLength - in/out parameter; contains the size of the input buffer. 
 * On successful read contains the number of bytes read from the underlying socket.
 */
OSStatus writeTLSConn(SSLConnectionRef connection, const void *data, size_t *dataLength)
{
    //NSLog(@"writeTLSConn called...\n");
    conn_t *tls_conn = (conn_t*) connection;
    int sockfd = tls_conn->sock_fd;
    
    size_t bytesToWrite = *dataLength;
    //CFStringRef errString;
    ssize_t status = 0;
    OSStatus osstatus = noErr;
    *dataLength = 0;
    tls_conn->want_write = 0;
    //NSLog(@"Bytes to write = %lu socket = %d", bytesToWrite, sockfd);

    while(1)
    {
        status = write(sockfd, data, bytesToWrite);
        if (status > 0)
        {
            *dataLength += status;
            if (bytesToWrite > *dataLength)
            {
                data += status;
            }
            else
            {
                osstatus = errSecSuccess;
                break;
            }
        }
        else if (0 == status)
        {
            osstatus = errSSLClosedGraceful;
            break;
        }
        else
        {
            if (EAGAIN == errno)
            {
                osstatus = errSSLWouldBlock;
                tls_conn->want_write = 1;
            }
            else
            {
                osstatus = errSecIO;
            }
            break;
        }
    }
    
    //errString = SecCopyErrorMessageString(osstatus, NULL);
    //NSLog(@"Number of bytes written = %lu, osstatus = %@", *dataLength, errString);
    return osstatus;
    
}


void performTLSHandshakeTest()
{
    struct sockaddr_in servaddr;
    
    int sock_fd = socket(AF_INET, SOCK_STREAM, 0);
    
    bzero(&servaddr, sizeof(servaddr));
    
    // assign IP, PORT
    servaddr.sin_family = AF_INET;
    servaddr.sin_addr.s_addr = inet_addr("172.217.31.206");
    servaddr.sin_port = htons(443);
    
    // TCP connect to server
    if (connect(sock_fd, (struct sockaddr*)&servaddr, sizeof(servaddr)) != 0) {
        printf("connection with the server failed...\n");
        exit(-1);
    }
    
    conn_t tls_conn;
    tls_conn.sock_fd = sock_fd;
    tls_conn.want_read = 0;
    tls_conn.want_write = 0;
    
    @autoreleasepool {
    //Create the TLS context object - we are on the client side and we would use TCP as transport
    SSLContextRef ssl_ctx = SSLCreateContext(kCFAllocatorDefault, kSSLClientSide, kSSLStreamType);
        //SSLContextRef ssl_ctx = NULL;
   //SSLNewContext(false, &ssl_ctx);
    
    //register the read/write IO callbacks
    SSLSetIOFuncs(ssl_ctx, readTLSConn, writeTLSConn);

    //Associate the TCP stream with the SSL Context created earlier
    SSLSetConnection(ssl_ctx, (SSLConnectionRef)&tls_conn);
    
    //SSLSetPeerDomainName(ssl_ctx,"a.r11.dot.dns.akasecure.net",strlen("a.r11.dot.dns.akasecure.net"));
    SSLSetPeerDomainName(ssl_ctx,"google.com",strlen("google.com"));
    
    SSLSetAllowsAnyRoot(ssl_ctx, true);
    
    OSStatus ssl_handshake_status = noErr;
    ssl_handshake_status = SSLHandshake(ssl_ctx);

    /*CFStringRef errString = SecCopyErrorMessageString(ssl_handshake_status, NULL);
    NSLog(@"Status of SSL Handshake:");
    NSLog((__bridge NSString *)errString);*/
    
    OSStatus ssl_close_status = SSLClose(ssl_ctx);
    //NSLog(@"Status of SSL Close:");
    /*errString = SecCopyErrorMessageString(ssl_close_status, NULL);
    NSLog((__bridge NSString *)errString);*/

    CFRelease(ssl_ctx);
    //SSLDisposeContext(ssl_ctx);
        
    // Close the underlying TCP socket
    close(tls_conn.sock_fd);
    }
}


int main(int argc, const char * argv[]) {
    
    int counter = 0;
    @autoreleasepool {
        while(1)
        {
            printf("Performing tls handshake...count: %d\n", counter);
            performTLSHandshakeTest();
            //sleep for 1 millisecond and perform another TLS handshake test
            usleep(1000);
            counter++;
        }
    }
    return 0;
}
---------------end program-----------------------

**Comment by wfurt** [comment]
Thanks for the update @dradhamohan. I'm planning to look at this after next public beta. The C repro is super useful as I plan to go back to 10.12 and use Valgrind to look for leaks. (unfortunately, it is not supported on current versions but it may be good enough to point at the problem)

**Comment by dradhamohan** [comment]
> Thanks for the update @dradhamohan. I'm planning to look at this after next public beta. The C repro is super useful as I plan to go back to 10.12 and use Valgrind to look for leaks. (unfortunately, it is not supported on current versions but it may be good enough to point at the problem)

**Comment by dradhamohan** [comment]
Sorry - I accidentally closed the ticket. Re-opened it. Was about to reply to your comment stating that the Apple leaks utility doesn't show any leak with the attached program though one can clearly see a memory increase. And we can see the same behavior across High-Sierra, Mojave and Catalina.

**Comment by wfurt** [comment]
I'm not familiar with that tool. Memory may not be shown as leak if there is reference to it. For example, session cache can grow but it is not leak because entries can be reached. I think we will need to figure out where the memory is used.
You can open issue with Apple - just in case there are some answers to it. With plain C it should be easier than giving them c#.

**Comment by karelz** [comment]
Triage: Happens in C as well, unlikely .NET Core direct problem - moving to Future. Still would be nice to have it understood and solved post 5.0.

**Comment by karelz** [comment]
This is now fixed in 6.0 in PR #41657
